### DISCUSSION TITLE
[Туториал] Как сделать поле "Статус" с переходом состояний из одного в другие

### DISCUSSION THREAD

**Discussion by simplepersonru** [totumonline/totum-ru-issues-and-discussions#418] [Как это сделать?] [answered]
Это не вопрос, а туториал

Иногда возникает потребность иметь поле в [расчетной в цикле](https://ru.docs.totum.online/tables#table-calc) таблице, которое будет изменяться согласно определенному бизнес-процессу. Рассмотрим на примере некоей "заявки"

Опишем все возможные бизнес-состояния нашей заявки [простой](https://ru.docs.totum.online/tables#table-simple) таблицей:
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/71624171/0d14fbe3-3ba0-46cd-8835-1d8944e43fcf)

**Название** и **Переход** - строки. **Варианты** - [множественный](https://ru.docs.totum.online/fields#fields-settings-multiple) селект, вот с таким кодом `= : selectRowListForSelect(table: $#ntn; field: 'perehod';)`. В **Варианты** для каждой строки самостоятельно добавляем переходы. Это те переходы, которые будут доступны, если выбран элемент с **Названием** из этой строки.

Далее создадим поле "статус" в таблице, где нужен этот бизнес-процесс. Код селекта поля "статус": 
```
= : listAdd(list: $selList; list: $currentSel;)
selList: selectRowListForSelect(table: 'status'; field: 'perehod'; where: 'id' = $ids)
	ids: select(table: 'status'; field: 'varianty'; where: 'id' = $#nfv;)
	
currentSel: selectRowListForSelect(table: 'status'; field: 'nazvanie'; where: 'id' = $#nfv)
```
Код берет все варианты переходов из **Варианты** и предлагает их к выбору, при этом в селекте присутствует и текущий выбранный вариант с обычным названием (не переходным).
Обязательно чекнуть вот эту галку:
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/71624171/27379285-9195-4c94-bf70-70e54eb02213)

В такой схеме нужно стартовое состояние, у нас это "Открыта". Напишем полю код действия при создании (здесь код для случая таблицы расчетной в цикле, поэтому используется $#nci.):
`= : set(table: $#ntn; field: 'h_status' = 1; where: 'id' = $#nci; log: false)`

Ниже пример работы с полем "статус". В этом примере была пройдена одна из возможных ветвей для бизнес-процесса заявки:

![asd](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/71624171/a73513de-0aa8-4eef-b0b7-8f41bb12c11e)

**Comment by ttm-alexeypolunin** [comment]
Я бы вот только не код действия при добавлении использовал, а Значение по умолчанию или Код только при добавлении (они одинаково отрабатывают). Тк код действия с set вызывает лишний пересчет всей таблицы.

Еще одно важное замечание для таких рассчитываемых Селектов — это безопасность. Если вам железно нужно, что бы на сервере было проверено, что пользователь вводит допустимое для него значение — нужно поставить еще галочку в Проверять введенные значения на сервере.

**Reply by ttm-alexeypolunin** [reply]
@acababaev напишите что-нибудь еще в этой ветке (хоть три точки), что бы я вас как answer смог поставить.

**Reply by simplepersonru** [reply, marked as answer]
Спасибо за ответ! Удивительно, но я тоже могу себе поставить marked as answer, вроде раньше не мог

**Comment by LostAlly** [comment]
Спасибо за решение, я его немного расширил следующим образом:
Думаю использовать состояния в разных таблицах, и не во всех нужны все состояния, плюс хотелось бы красить состояния  в одинаковые цвета не дублируя
Добавил поля цвет(тип строка) и таблицы(селект с множественным выбором и прописанным статически списком таблиц)
При формировании списка добавил условие проверки имени таблицы и заливка ячейки берется из таблицы состояния.
![image](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/5542325/22d2fa23-aaf7-48c5-bdc9-c1c4bb0f9dc3)

Я еще изучаю тотум, не дошел до разукрашивания селекта, было бы еще удобней.

Если нужно могу выложить более подробно код который был добавлен, но там сущие пустяки.

И было бы здорово, чтобы в тотуме была системная таблица с системными цветами, чтобы ее можно было просто подключать в свои таблицы.

P.S. с включенной галочкой - "проверять на сервере" значения не сохраняются.

**Reply by ttm-alexeypolunin** [reply]
P.S. с включенной галочкой - "проверять на сервере" значения не сохраняются.

Если **отключен Индивидуальный Селект** для каждой строки и в коде селекта есть условия (например зависимость допуст значений от id в тек таблице), то **Селект будет рассчитан для одной случайной строки.**

Таким образом в других строках значения в выпадающем списке будут такие-же. В этом случае может быть ситуация, что там будут в том числе недопустимые значения, которые и будут отклонены при попытке их выбрать на стороне сервера.

**Reply by konditerov** [reply]
> 
> Я еще изучаю тотум, не дошел до разукрашивания селекта, было бы еще удобней.
> 
> Если нужно могу выложить более подробно код который был добавлен, но там сущие пустяки.
> 

Подскажите, как вы раскрасили селект цветом из другой таблицы..

upd:
`f1=: setFormat(background: $color)
color: select(table: 'statusy_dostavki'; field: 'cvet'; where: 'id' = $#nfv;)`
